### PR TITLE
SCIM: Enable SCIM for existing users in case of PATCH or PUT

### PR DESCRIPTION
**Pull request by vdavid**
- Fixes https://github.com/sourcegraph/sourcegraph/issues/48441

**Context:** Some users are **already in the DB** with the same email address as in the IdP. GET requests return non-SCIM-enabled users, so IdPs can know that these users already exist, despite them not having a SCIM entry in `user_external_accounts`. `CREATE` requests correctly respond with `409` in case the IdP tries to create an existing user. In case of `PATCH` and `PUT` requests, we expect these to change user data _and_ enable SCIM for these users.

**Problem:** `PATCH` and `PUT` requests return a valid response, but they fail to change the user in the DB, and they leave `user_external_accounts` intact.

**Solution:** With the changes in this PR, we'll notice if there is no entry in `user_external_accounts` when modifying/replacing the user, and create one.

## Test plan

Tested with Postman, both with SCIM-enabled users (to avoid a regression) and non-SCIM-enabled users (to test the new feature) and it worked smoothly.

### CODE REVIEW

File: internal/database/external_accounts.go [hunk 1]
```diff
@@ -163,27 +164,39 @@ RETURNING user_id
...
 WHERE
-	user_id = $1
-AND service_type = $2
-AND service_id = $3
+	user_id = %s
+AND service_type = %s
+AND service_id = %s
 AND deleted_at IS NULL
-`, userID, "scim", "scim", accountID, encryptedAuthData, encryptedAccountData, keyID)
+`, accountID, encryptedAuthData, encryptedAccountData, keyID, userID, "scim", "scim"))
```
Comment (vdavid) [inline internal/database/external_accounts.go:187]: I also converted this to `sqlf` as a byproduct

File: enterprise/internal/scim/user.go [hunk 2]
```diff
@@ -86,11 +86,11 @@ func createUserResourceType(userResourceHandler *UserResourceHandler) scim.Resou
 }
 
 // updateUser updates a user in the database. This is meant to be used in a transaction.
-func updateUser(ctx context.Context, db database.DB, oldUser *types.UserForSCIM, attributes scim.ResourceAttributes) (err error) {
+func updateUser(ctx context.Context, tx database.DB, oldUser *types.UserForSCIM, attributes scim.ResourceAttributes) (err error) {
```
Comment (vdavid) [inline enterprise/internal/scim/user.go:89]: This function has a variable rename unrelated to this PR; sorry, I just sneaked it in.
The only "real" change is the function name in line 114.

File: internal/database/external_accounts.go [hunk 3]
```diff
@@ -163,27 +164,39 @@ RETURNING user_id
...
+	if err != nil {
+		return
+	}
+	rowsAffected, err := res.RowsAffected()
+	if err != nil {
+		return
+	}
+
+	if rowsAffected == 0 {
+		return s.Insert(ctx, userID, extsvc.AccountSpec{ServiceType: "scim", ServiceID: "scim", AccountID: accountID}, data)
```
Comment (vdavid) [inline internal/database/external_accounts.go:197]: This is the core of the change. Call `Insert` if we found no row

File: internal/database/external_accounts.go [hunk 4]
```diff
@@ -78,10 +78,11 @@ type UserExternalAccountsStore interface {
 	// CreateUserAndSave for that.
 	LookupUserAndSave(ctx context.Context, spec extsvc.AccountSpec, data extsvc.AccountData) (userID int32, err error)
 
-	// UpdateSCIMData updates the external account data for the given user's SCIM account.
+	// UpdateOrInsertSCIMData updates the external account data for the given user's SCIM account.
```
Comment (chwarwick) [inline internal/database/external_accounts.go:81]: `UpsertSCIMData`?

File: internal/database/external_accounts.go [hunk 5]
```diff
@@ -163,27 +164,39 @@ RETURNING user_id
...
 
-func (s *userExternalAccountsStore) UpdateSCIMData(ctx context.Context, userID int32, accountID string, data extsvc.AccountData) (err error) {
+func (s *userExternalAccountsStore) UpdateOrInsertSCIMData(ctx context.Context, userID int32, accountID string, data extsvc.AccountData) (err error) {
 	encryptedAuthData, encryptedAccountData, keyID, err := s.encryptData(ctx, data)
 	if err != nil {
 		return
 	}
 
-	_, err = s.Handle().ExecContext(ctx, `
+	res, err := s.ExecResult(ctx, sqlf.Sprintf(`
```
Comment (chwarwick) [inline internal/database/external_accounts.go:173]: I guess the alternative would be to rewrite this as an upsert statement, but my assumption is that the insert case is expected to be very infrequent so I think this makes sense to leave as an update and then only do the insert if neccessary.

**Review by chwarwick — APPROVED** [review]
Looks good.

File: internal/database/external_accounts.go (hunk 4 above)
Comment (vdavid) [reply internal/database/external_accounts.go]: Yup, better name, renamed it to that

**Pull request merged**